spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/CassandraRow.scala

```diff
@@ -86,18 +153,37 @@ object CassandraRow {
     * the newly created `CassandraRow`, but it is not used to fetch data from
     * the input `Row` in order to improve performance. Fetching column values by name is much
     * slower than fetching by index. */
-  def fromJavaDriverRow(row: Row, columnNames: Array[String]): CassandraRow = {
-    val data = new Array[Object](columnNames.length)
-    for (i <- columnNames.indices)
-      data(i) = GettableData.get(row, i)
-    new CassandraRow(columnNames, data)
+  def fromJavaDriverRow(row: Row, metaData: CassandraRowMetadata): CassandraRow = {
+    val length = metaData.columnNames.length
+    var i = 0
+    val data = new Array[Object](length)
+
+    // Here we use a mutable while loop for performance reasons, scala for loops are
+    // converted into range.foreach() and the JVM is unable to inline the foreach closure.
+    // 'match' is replaced with 'if' for the same reason.
+    // It is also out of the loop for performance.
+    if (metaData.codecs == null) {
+      //that should not happen in production, but just in case
+      while (i < length) {
+        data(i) = GettableData.get(row, i)
+        i += 1
+      }
+    }
+    else {
+      while (i < length) {
+        data(i) = GettableData.get(row, i, metaData.codecs(i))
+        i += 1
+      }
+    }
+    new CassandraRow(metaData, data)
```

Comment:
I have rewiten that block to java style to remove closure object creation for each row